store origin device of node_ip entries in custom_fields.seen

rc9000 opened the pull request:
Hello @ollyg & team! Here's a first version of what's discussed in #1053 and on the IRC.
If we agree on this implementation I'll make a second patch that shows this information in the UI.

ollyg commented:
Good approach! My initial thoughts are below. These are just thoughts and not requirement to do anything different, and some is just thinking out loud the options. Let me know what you think.
- do we want to stomp on the "seen" slot of user-defined custom fields? we'd have to document it and guard against a user setting/messing with it. options are to use something other than custom_fields (like system_fields) or else to prefix our fields with netdisco_ or similar and document/protect that (easier?).
- the implementation makes a dictionary and updates the router entry with the timestamp. just noting that this is not a "history" which would be a list where entries are added when the router IP changes, or latest is updated (so you can go routerA, routerB, routerA again, etc). However as Netdisco doesn't support L3 virtualisation I think your implementation is safer, otherwise we could have an explosion if two routers flip-flop entries.
- I'd like to find a more efficient database update algorithm than update_or_new then insert with the branch, but it's not essential.

rc9000 commented:
> 
>     * do we want to stomp on the "seen" slot of user-defined custom fields? we'd have to document it and guard against a user setting/messing with it. options are to use something other than custom_fields (like system_fields) or else to prefix our fields with netdisco_ or similar and document/protect that (easier?).

I like the prefix idea and suspect similar cases might come up in other tables, let's just say netdisco_ keys are  implementation namespace in all custom_fields columns and should not be used? Or maybe better `_netdisco_` since this is a convention people might be familiar with already.

> 
>     * the implementation makes a dictionary and updates the router entry with the timestamp. just noting that this is not a "history" which would be a list where entries are added when the router IP changes, or latest is updated 

True. It's nice that it's currently a self-managed structure that will not grow infinitely in size. Maybe as a one-step inbetween compromise we could have first/last fields like all other historic data does. 

> 
>     * I'd like to find a more efficient database update algorithm than update_or_new then insert with the branch, but it's not essential.

Yes that bugged me too and took basically all of the time that went into this PR, but I didn't find a better way :)

rc9000 commented:
ok it's now `_netdisco_seen_in_arp` and with time_first/time_last. But updating deep json structures like this is pretty horrid with the jsonb_ functions. I gave up around

```
    $row->set_column(
        custom_fields => \[ qq{ jsonb_set( custom_fields,
                    '{_netdisco_seen_in_arp, ?}',
                    jsonb_build_object( ?::text, coalesce( custom_fields#>>'{_netdisco_seen_in_arp, ?}'::text[],
                            'null')::jsonb || jsonb_build_object( 'time_last', to_jsonb(to_timestamp(?))
                        )), true
                )
            },$device_ip, $device_ip, $device_ip, $now ]
    ) if $device_ip;
```

which also would have needed a similar jsonb_build_object counterpart for the insert case, and switched to JSON::XS `(en|de)code_json`. 

Roundtripping and overwriting the whole jsonb each time is probably less efficient, but way more readable. Shouldn't also get too big unless you run into #1185 which gave me millions of `_netdisco_seen_in_arp`in no time. I started to doubt the network design :)

Also added an index for the jsonb like all the other tables have.
